libethereum/EthereumHost.cpp

```diff
@@ -54,36 +54,41 @@ namespace
 class EthereumPeerObserver: public EthereumPeerObserverFace
 {
 public:
-	EthereumPeerObserver(BlockChainSync& _sync, RecursiveMutex& _syncMutex, TransactionQueue& _tq): m_sync(_sync), m_syncMutex(_syncMutex), m_tq(_tq) {}
+	EthereumPeerObserver(shared_ptr<BlockChainSync> _sync, weak_ptr<TransactionQueue> _tq): m_sync(_sync), m_tq(_tq) {}
 
 	void onPeerStatus(std::shared_ptr<EthereumPeer> _peer) override
 	{
-		RecursiveGuard l(m_syncMutex);
+		if (!m_sync)
```

Comment:
Why do you need these checks? Isn't it always non-null?

Reply:
Without these checks, I see a segmentation fault.

Reply:
`AddressSanitizer` reports an access to zero-page, but cannot show the source line.

With `gdb` I get more clue
```
$ gdb --args eth/eth --ropsten --kill --pin --peerset "required:20c9ad97c081d63397d7b685a412227a40e23c8bdc6688c6f37e97cfbc22d2b4d1db1510d8f61e6a8866ad7f0e17c02b14182d37ea7c3c8b9c2683aeb6b733a1@52.169.14.227:30303 required:6ce05930c72abc632c58e2e4324f7c7ea478cec0ed4fa2528982cf34483094e9cbc9216e7aa349691242576d552a2a56aaeae426c5303ded677ce455ba1acd9d@13.84.180.240:30303" --port 30304

(gdb) run
Thread 31 "p2p" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe5ee3700 (LWP 6139)]
__GI___pthread_mutex_lock (mutex=0x28) at ../nptl/pthread_mutex_lock.c:67
67      ../nptl/pthread_mutex_lock.c: No such file or directory.
#1  0x00005555557707ef in __gthread_mutex_lock (__mutex=0x28)
    at /usr/include/x86_64-linux-gnu/c++/7/bits/gthr-default.h:748
#2  __gthread_recursive_mutex_lock (__mutex=0x28)
    at /usr/include/x86_64-linux-gnu/c++/7/bits/gthr-default.h:810
#3  std::recursive_mutex::lock (this=0x28) at /usr/include/c++/7/mutex:107
#4  std::lock_guard<std::recursive_mutex>::lock_guard (__m=..., this=<synthetic pointer>)
    at /usr/include/c++/7/bits/std_mutex.h:162
#5  dev::eth::BlockChainSync::onPeerAborting (this=0x0)
    at /home/yh/src/cpp-ethereum/libethereum/BlockChainSync.cpp:947
#6  0x00005555556f11ac in (anonymous namespace)::EthereumPeerObserver::onPeerAborting (
    this=<optimized out>) at /home/yh/src/cpp-ethereum/libethereum/EthereumHost.cpp:87
#7  0x00005555557003be in dev::eth::EthereumPeer::abortSync (this=this@entry=0x7fffd80682e0)
    at /home/yh/src/cpp-ethereum/libethereum/EthereumPeer.cpp:113
#6  0x00005555556f11ac in (anonymous namespace)::EthereumPeerObserver::onPeerAborting (
    this=<optimized out>) at /home/yh/src/cpp-ethereum/libethereum/EthereumHost.cpp:87
87                              m_sync->onPeerAborting();
```

Reply:
@gumb0 told me that I left `EthereumPeerObserver` unaware of the `BlockChainSync` object.  If I fix this, I shouldn't need `if(m_sync)` checks.